sources/api/apiclient/src/update.rs

```diff
@@ -0,0 +1,393 @@
+`most_recent_command` structure giving details about the time, type, and result of the last issued
+command.  So, this is our basic process:
+* Check initial status, saving the timestamp.
+* Issue our request.
+* Poll the status API (with timeout and failure checks) until the timestamp changes.
+* Confirm that the status shows that our requested command was successful.
+
+Note: there's some potential for conflict with other callers, even with the API locking.  If
+someone else issues a command after ours completes, but before we can check, the status could be
+lost.  The update API's state machine ensures we only move in expected directions, and we don't
+want to reinvent its logic here, or be too strict about timing.  We can just time out, or perhaps
+sync up if their request was the same.  If it becomes a problem, we could perhaps use something
+like transactions, or timed lock files, to avoid it.
+*/
+
+use super::{raw_request, raw_request_unchecked};
+use http::StatusCode;
+use log::{debug, info, trace, warn};
+use snafu::{ensure, ResultExt};
+use std::path::Path;
+use std::thread;
+use std::time::Duration;
+
+/// Refresh the list of available updates and return the current status.
+pub async fn check<P>(socket_path: P) -> Result<String>
+where
+    P: AsRef<Path>,
+{
+    info!("Refreshing updates...");
+    let (_body, status) = wait_request(
+        socket_path,
+        "/actions/refresh-updates",
+        "POST",
+        None,
+        "refresh",
+        &WaitPolicy::new(Duration::from_millis(100), 10 * 10),
+    )
+    .await?;
+
+    Ok(status)
+}
+
+/// Checks whether an update is required given the current update status, and informs the user
+/// about the specific state.
+// Note: we shouldn't reimplement the update API's state tracking or knowledge here, but giving
+// a useful message rather than "failed to prepare update" in common scenarios is critical.
+pub fn required(check_output: &str) -> bool {
+    let state =
+        response_field(&["update_state"], check_output).unwrap_or_else(|| "unknown".to_string());
+    match state.as_ref() {
+        "Idle" => {
+            info!("No updates available.");
+            false
+        }
+        "Ready" => {
+            info!("Update already applied; reboot for it to take effect, or request a cancel.");
+            false
+        }
+        _ => {
+            debug!("Saw state '{}', assuming update is required", state);
+            true
+        }
+    }
+}
+
+/// Applies the update shown as selected in the output of check(), and makes it active.
+pub async fn apply<P>(socket_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    info!("Downloading and applying update to disk...");
+    let (_body, _status) = wait_request(
+        &socket_path,
+        "/actions/prepare-update",
+        "POST",
+        None,
+        "prepare",
+        &WaitPolicy::new(Duration::from_millis(500), 2 * 60 * 10),
```

Comment:
I wonder if 10 minutes is enough. If it's not enough time for the host due to its network condition, then the user is out of luck. Maybe this would be the case for raspberry pi's/IoT devices sitting in networks with poor bandwidth? 

Can we make this configurable somehow? If not, can we increase to maybe 20 minutes? I don't see the harm in letting this run longer, the user always have the option of instituting their own timeouts or ctrl-c out of it if this becomes too long.

Reply:
I'm conflicted here.  A 10 minute timeout, with our current ~239MB release size, means a requirement of ~409KB/s.  I suspect you'd have a hard time making any cloud variant fail that requirement.  We don't support local/small devices yet, but when we do, I question whether longer timeouts is the right answer here.  It feels more like thar-be-updates / the update API needs a remodel (so it's not just locked for an hour+) or, at a minimum, apiclient is (much further) remodeled to use websockets or local state or something.

So, while a timeout argument wouldn't be hard to add, I don't think it reflects the reality of the world now, nor does it reflect where I think we'd want to go in the future to handle that type of constraint.  Does that make sense?  Do we have other options?

(The user can always check status after apiclient times out, with the current code, and issue another request to move things forward when the lock opens...)

Reply:
While my spidey senses are telling me we may run into this issue sometime later, I agree with @tjkirch.  Given our current release size and the meager network requirements to make it all happen, I think leaving the current timeout is fine.  Something something YAGNI...

Reply:
Yeah, I agree. I suppose it's too early to say how updates on smaller devices would look like.

Can we add a comment above the WaitPolicy to explain how long the total wait is? At a glance it might not be obvious.
```suggestion
        // This equates to a 10 minute timeout policy.
        &WaitPolicy::new(Duration::from_millis(500), 2 * 60 * 10),
```
Or we can document these values elsewhere, maybe in the README?

Reply:
While my spidey senses are telling me we may run into this issue sometime later, I agree with @tjkirch.  Given our current release size and the meager network requirements to make it all happen, I think leaving the current timeout is fine.  Something something YAGNI...

Reply:
> Or we can document these values elsewhere, maybe in the README?

I'm not sure what you mean here, which README?  The struct and the function that uses it are private, so I don't think they should go into docs, beyond their own docstrings (which do explain the values).